### ISSUE TITLE
Is decoder output returning?

### ISSUE THREAD

**Issue by Akseli-Ilmanen** [Wangjhdeveloper/CETNet#2]
Hi,

Really cool model!

I was wondering whether there is a typo in the code below. If the coder should output in `def_forward` of `MyTransformer` should the lines below not says `for i, cf in enumerate(ce_features):`, this way the model will loop through the decoder iterations? Maybe I am missing something?

```
    def forward(self, x, mask):
        out, feature, ce_features = self.encoder(x, mask)

        outputs = out.unsqueeze(0)

        i = 0
        for cf in ce_features:
            if i != 0:
                out, feature = self.decoder(F.softmax(out, dim=1) * mask[:, 0:1, :], cf * mask[:, 0:1, :], mask)
            i = + 1
            outputs = torch.cat((outputs, out.unsqueeze(0)), dim=0)
        features = nn.functional.normalize(feature).unsqueeze(0)
       
        return outputs, features
```

https://github.com/Wangjhdeveloper/CETNet/blob/c9c6a2f7dae7a4043203eaf942d4e075259354f8/model.py#L360C1-L373C33

All the best
Akseli

**Comment by Akseli-Ilmanen** [comment]
Another question, do we need the `if i != 0:`, is that equivalent to not passing the outputs of the first encoder layer to the decoder?

```
  def forward(self, x, mask):
      out, feature, ce_features = self.encoder(x, mask)

      outputs = out.unsqueeze(0)

      # i = 0
      for i, cf in enumerate(ce_features):
          # if i != 0:
          out, feature = self.decoder(F.softmax(out, dim=1) * mask[:, 0:1, :], cf * mask[:, 0:1, :], mask)

          outputs = torch.cat((outputs, out.unsqueeze(0)), dim=0)
      features = nn.functional.normalize(feature).unsqueeze(0)


      return outputs, features

```

**Comment by Akseli-Ilmanen** [comment]
Wait nevermind, in my code, I must have accidently deleted the `i = + 1` (althouth this should be `i += 1`) that's why it was not working. Sorry! Will close this issue.

**Comment by Akseli-Ilmanen** [comment]
Hi again,

So, I now corrected the `i = + 1` to `i += 1`, and retrained the model. Problematically, model training took much longer (around 30x slower), and also the test accuracy went down by a lot. 

Is there a change one can make to the code to include the decoder as described in the paper?

Best
Akseli
